core/controller/src/main/scala/org/apache/openwhisk/core/controller/Packages.scala

```diff
@@ -121,29 +122,42 @@ trait WhiskPackagesApi extends WhiskCollectionAPI with ReferencedEntities {
    * - 500 Internal Server Error
    */
   override def remove(user: Identity, entityName: FullyQualifiedEntityName)(implicit transid: TransactionId) = {
-    deleteEntity(
-      WhiskPackage,
-      entityStore,
-      entityName.toDocId,
-      (wp: WhiskPackage) => {
-        wp.binding map {
-          // this is a binding, it is safe to remove
-          _ =>
-            Future.successful({})
-        } getOrElse {
-          // may only delete a package if all its ingredients are deleted already
-          WhiskAction
-            .listCollectionInNamespace(entityStore, wp.namespace.addPath(wp.name), skip = 0, limit = 0) flatMap {
-            case Left(list) if (list.size != 0) =>
-              Future failed {
-                RejectRequest(
-                  Conflict,
-                  s"Package not empty (contains ${list.size} ${if (list.size == 1) "entity" else "entities"})")
-              }
-            case _ => Future.successful({})
+    parameter('force ? false) { force =>
+      deleteEntity(
+        WhiskPackage,
+        entityStore,
+        entityName.toDocId,
+        (wp: WhiskPackage) => {
+          wp.binding map {
+            // this is a binding, it is safe to remove
+            _ =>
+              Future.successful({})
+          } getOrElse {
+            // may only delete a package if all its ingredients are deleted already or force flag is set
+            WhiskAction
+              .listCollectionInNamespace(entityStore, wp.namespace.addPath(wp.name), includeDocs = true, skip = 0, limit = 0) flatMap {
+              case Right(list) if list.nonEmpty =>
+                if (force) {
+                  Future sequence {
+                    list.map(action => {
+                      WhiskAction.get(entityStore, wp.fullyQualifiedName(false).add(action.fullyQualifiedName(false).name).toDocId) flatMap { actionWithRevision =>
```

Comment:
This LGMT but there are some failure modes to consider or at least document:
- failures to get
- failures to delete
- concurrent modifications

(We don't have a global lock on CRUD ops in namespace.)
So if any of those futures fail, should the future sequence also fail and report an error?

Reply:
Yea good point. From what I can tell with other entity deletions, they all go through the `deleteEntity` function which will get the document first to get the revision so I think failure to get is the same as any other deletion.

For failure to delete, some of the list of actions can fail to delete and the future sequence will fail if a single action fails to delete. And since the actions are deleted as a part of the confirmation phase of the package delete, the package won't delete unless they all succeed. I just need to go back and make sure when an action fails to delete it propagates the correct api failure or I recover here to transform it to the correct api exception. Should be clear to the client that if they get a failure for one of the actions failing to delete, they need to redo the request.

For concurrent modifications I wasn't really thinking about. What's the consideration I should make there with this?

Reply:
concurrent mods would manifest as failure to delete in this case - so addressing the get/del failure by bubbling up the error should suffice.

Reply:
So I'm a little stuck on what the best way to handle the failures of the get and deletes of individual actions are. Since the future sequence is a part of the confirm phase of the `deleteEntity` function, if any of them fail; it will be caught by the `deleteEntity` failure handling. That might be good enough for example if the `get` fails on one of the actions because it was concurrently deleted while the request was in flight, it gets a `DocumentNotFound` exception and will be caught and resolve the api with a `NotFound` error as expected. The error message will just be entity not found so not specific as to whether the action failed to delete within it or the package itself. I don't know what the thoughts are on whether that's sufficient, I tried to come up with something to combine some code for the failure handling to reuse but didn't really come up with anything good.

I personally don't think it's a big issue as long as it's documented that the returned error may not be specific to the package or an action within it, but curious what you think.